File: gen_coverage/fastcov.py


def distillFunction(function_raw, functions):
    function_name   = function_raw["name"]
    # NOTE: need to explicitly cast all counts coming from gcov to int - this is because gcov's json library
    # will pass as scientific notation (i.e. 12+e45)
    start_line      = int(function_raw["start_line"])
    execution_count = int(function_raw["execution_count"])
    if function_name not in functions:
        functions[function_name] = {
            "start_line": start_line,
            "execution_count": execution_count
        }
    else:
        functions[function_name]["execution_count"] += execution_count

def emptyBranchSet(branch1, branch2):
    return (branch1["count"] == 0 and branch2["count"] == 0)

def matchingBranchSet(branch1, branch2):
    return (branch1["count"] == branch2["count"])

def filterExceptionalBranches(branches):
    filtered_branches = []
    exception_branch = False
    for i in range(0, len(branches), 2):
        if i+1 >= len(branches):
            filtered_branches.append(branches[i])
            break

        # Filter exceptional branch noise
        if branches[i+1]["throw"]:
            exception_branch = True
            continue

        # Filter initializer list noise
        if exception_branch and emptyBranchSet(branches[i], branches[i+1]) and len(filtered_branches) >= 2 and matchingBranchSet(filtered_branches[-1], filtered_branches[-2]):
            return []

        filtered_branches.append(branches[i])
        filtered_branches.append(branches[i+1])

    return filtered_branches

def distillLine(line_raw, lines, branches, include_exceptional_branches):
    line_number = int(line_raw["line_number"])
    count       = int(line_raw["count"])
    if count <  0:
        if "function_name" in line_raw:
            print("WARN: Ignoring negative count found in '%s'.", line_raw["function_name"])
        else:
            print("WARN: Ignoring negative count.")
        count = 0

    if line_number not in lines:
        lines[line_number] = count
    else:
        lines[line_number] += count

    # Filter out exceptional branches by default unless requested otherwise
    if not include_exceptional_branches:
        line_raw["branches"] = filterExceptionalBranches(line_raw["branches"])

    # Increment all branch counts
    for i, branch in enumerate(line_raw["branches"]):
        if line_number not in branches:
            branches[line_number] = []
        blen = len(branches[line_number])
        glen = len(line_raw["branches"])
        if blen < glen:
            branches[line_number] += [0] * (glen - blen)
        branches[line_number][i] += int(branch["count"])

def distillSource(source_raw, sources, test_name, include_exceptional_branches):
    source_name = source_raw["file_abs"]
    if source_name not in sources:
        sources[source_name] = {
            test_name: {
                "functions": {},
                "branches": {},
                "lines": {}
            }
        }

    for function in source_raw["functions"]:
        distillFunction(function, sources[source_name][test_name]["functions"])

    for line in source_raw["lines"]:
        distillLine(line, sources[source_name][test_name]["lines"], sources[source_name][test_name]["branches"], include_exceptional_branches)

File: rust/src/fastcov.rs
use serde_json::Value;
use std::collections::HashMap;

pub fn distill_function(function_raw: &Value, functions: &mut HashMap<String, HashMap<String, i64>>) {
    let function_name = function_raw["name"].as_str().unwrap();
    let start_line = function_raw["start_line"].as_i64().unwrap();
    let execution_count = function_raw["execution_count"].as_i64().unwrap();

    if !functions.contains_key(function_name) {
        let mut func_info = HashMap::new();
        func_info.insert("start_line".to_string(), start_line);
        func_info.insert("run_count".to_string(), execution_count);
        functions.insert(function_name.to_string(), func_info);
    } else {
        let func_info = functions.get_mut(function_name).unwrap();
        *func_info.get_mut("run_count").unwrap() += execution_count;
    }
}

pub fn empty_branch_set(branch1: &Value, branch2: &Value) -> bool {
    branch1["count"].as_i64().unwrap() == 0 && branch2["count"].as_i64().unwrap() == 0
}

pub fn matching_branch_set(branch1: &Value, branch2: &Value) -> bool {
    branch1["count"].as_i64().unwrap() == branch2["count"].as_i64().unwrap()
}

pub fn filter_exceptional_branches(branches: &Vec<Value>) -> Vec<Value> {
    let mut filtered_branches = Vec::new();
    let mut exception_branch = false;

    for i in (0..branches.len()).step_by(2) {
        if i + 1 >= branches.len() {
            filtered_branches.push(branches[i].clone());
            break;
        }

        if branches[i + 1]["throw"].as_bool().unwrap_or(false) {
            exception_branch = true;
            continue;
        }

        if exception_branch && empty_branch_set(&branches[i], &branches[i + 1])
            && filtered_branches.len() >= 2
            && matching_branch_set(&filtered_branches[filtered_branches.len() - 1], &filtered_branches[filtered_branches.len() - 2])
        {
            return Vec::new();
        }

        filtered_branches.push(branches[i].clone());
        filtered_branches.push(branches[i + 1].clone());
    }

    filtered_branches
}

pub fn distill_line(
    line_raw: &Value,
    lines: &mut HashMap<i64, i64>,
    branches: &mut HashMap<i64, Vec<i64>>,
) {
    let line_number = line_raw["line_number"].as_i64().unwrap();
    let mut count = line_raw["count"].as_i64().unwrap();

    if count < 0 {
        if let Some(function_name) = line_raw.get("function_name") {
            println!(
                "WARN: Ignoring negative count found in '{}'.",
                function_name.as_str().unwrap()
            );
        } else {
            println!("WARN: Ignoring negative count.");
        }
        count = 0;
    }

    *lines.entry(line_number).or_insert(0) += count;

    let line_branches = line_raw["branches"].as_array().unwrap();
    for (i, branch) in line_branches.iter().enumerate() {
        let branch_count = branch["count"].as_i64().unwrap();
        let branch_entry = branches.entry(line_number).or_insert_with(|| vec![0; line_branches.len()]);
        if branch_entry.len() < line_branches.len() {
            branch_entry.resize(line_branches.len(), 0);
        }
        branch_entry[i] += branch_count;
    }
}

pub fn distill_source(
    source_raw: &Value,
    sources: &mut HashMap<String, HashMap<String, SourceData>>,
    test_name: &str,
) {
    let source_name = source_raw["file_abs"].as_str().unwrap().to_string();

    // Ensure the source and test_name entry exists
    if !sources.contains_key(&source_name) {
        let mut test_entry = HashMap::new();
        let source_data = SourceData {
            functions: HashMap::new(),
            branches: HashMap::new(),
            lines: HashMap::new(),
        };
        test_entry.insert(test_name.to_string(), source_data);
        sources.insert(source_name.clone(), test_entry);
    }

    let test_entry = sources.get_mut(&source_name).unwrap();
    let source_data = test_entry.get_mut(test_name).unwrap();

    // Handle function data
    for function in source_raw["functions"].as_array().unwrap() {
        distill_function(function, &mut source_data.functions);  // Correct type
    }

    // Handle line and branch data
    for line in source_raw["lines"].as_array().unwrap() {
        distill_line(line, &mut source_data.lines, &mut source_data.branches);
    }
}

#[derive(Debug)]
pub struct SourceData {
    pub functions: HashMap<String, HashMap<String, i64>>,
    pub branches: HashMap<i64, Vec<i64>>,
    pub lines: HashMap<i64, i64>,
}
